src/System.Management.Automation/engine/regex.cs

```diff
@@ -278,6 +278,64 @@ public static bool ContainsWildcardCharacters(string pattern)
             return result;
         }
 
+        /// FIXME
+        /// For invalid wildcard pattern, it currently only check if there is
+        /// unclosed bracket. Need to find a better solution and rewrite this.
+        ///
+        /// <summary>
+        /// Checks to see if the given string might contains wildcard pattern.
+        /// </summary>
+        /// <param name="pattern">
+        /// String which needs to be checked
+        /// </param>
+        /// <returns>
+        /// true if the string does not contain invalid wildcard pattern,
+        /// false otherwise.
+        /// </returns>
+        /// <remarks>
+        /// Currently string contains { '*', '?' } or both { '[', ']' } is
+        /// considered to be properbly a wildcard pattern and
+        /// '`' is the escape character.
+        /// </remarks>
+        public static bool ContainsValidWildcardPattern(string pattern)
+        {
+            if (string.IsNullOrEmpty(pattern))
+            {
+                return false;
+            }
+
+            bool result = false;
+            bool opened = false;
+
+            for (int index = 0; index < pattern.Length; ++index)
+            {
+                switch (pattern[index])
+                {
+                    case '*':
+                    case '?':
+                        result = true;
+                        break;
+
+                    case '[':
+                        result = true;
+                        opened = true;
+                        break;
+
+                    case ']':
+                        result = true;
+                        opened = false;
+                        break;
+
+                    // If it is an escape character then advance past
+                    // the next character
+                    case escapeChar:
+                        ++index;
+                        break;
+                }
+            }
+            return result && !opened;
```

Comment:
Please correct me if I'm wrong, but this will return `true` for a string like `"]abc"`.
For this PR, let's make this method able to discover unpaired `[` and `]` and incorrect order of `[` and `]`.

```c#
public static bool ContainsValidWildcardPattern(string pattern)
{
    if (string.IsNullOrEmpty(pattern))
    {
        return false;
    }

    bool result = false;
    int openBracketCount = 0;

    for (int index = 0; index < pattern.Length; ++index)
    {
        switch (pattern[index])
        {
            case '*':
            case '?':
                result = true;
                break;

            case '[':
                result = true;
                openBracketCount++;
                break;

            case ']':
                result = true;
                openBracketCount--;
                break;

            // If it is an escape character then advance past
            // the next character
            case escapeChar:
                index++;
                break;
        }

        if (openBracketCount < 0)
        {
            // Found ']' appears before '[', so it's not a valid wildcard pattern.
            return false;
        }
    }
    return result && openBracketCount == 0;
}
```

Reply:
Is the intent of this method to return `false` if no wildcards are used at all? (i.e. the string is "valid", but contains no wildcard usage). Currently it will return false for the input `"string"`. It seems simpler to me for the method to just check that any wildcard usage is valid. That way the monotonic `result` variable can be thrown out.

Also the string `"duck["` will cause an `IndexOutOfRangeException` here.

Possible implementation:
```csharp
internal static bool ContainsInvalidWildcardUsage(string pattern)
{
    // Empty patterns are valid
    if (pattern == null)
    {
        return true; // Possibly throw an exception -- depends on semantic interpretation
    }

    int bracketDepth = 0;
    for (int i = 0; i < pattern.Length; i++)
    {
        switch (pattern[i])
        {
            case escapeChar:
                 i++; // Skip the next character
                 continue;

            case '[':
                if (i + 1 < pattern.Length && pattern[i + 1] == ']')
                {
                    return true;
                }
                bracketDepth++;
                continue;

            case ']':
                bracketDepth--;
                if (bracketDepth < 0)
                {
                     return true;
                }
                continue;
        }
    }

    return bracketDepth == 0;
}
```

Reply:
Then you can get the `ContainsValidWildcardPattern` behaviour with:
```csharp
internal static bool ContainsValidWildcardPattern(string pattern)
{
    if (ContainsInvalidWildcardUsage(pattern))
    {
        return false;
    }

    foreach (char c in pattern)
    {
        case '[':
        case ']':
        case '*':
        case '?':
              return true;
    }

    return false;
}
```

Reply:
> Is the intent of this method to return false if no wildcards are used at all?

Yes, the purpose is to have an relatively accurate answer for "Is this string contains any wildcard pattern".

> Also the string `"duck["` will cause an IndexOutOfRangeException here.

Good catch! Length needs to be checked against 'index + 1' in `case '[':`.

> Then you can get the `ContainsValidWildcardPattern` behaviour with:

There are 2 concerns with this:
- We don't really need `ContainsInvalidWildcardUsage` for the issue in discussion.
- You may need to loop through the string twice in worse case, and 1.5 times in average case.

Reply:
Yes, the second loop is the price of composability -- and the average pattern is not going to be long, so a pretty low price. I would argue that the fact that we don't already have a method that can identify a string with invalid wildcards in it is surprising to say the least.

Reply:
> if (pattern[index + 1] == ']')

Boundary overflow.

Reply:
> I would argue that the fact that we don't already have a method that can identify a string with invalid wildcards in it is surprising to say the least.

I guess that's because we haven't got a need for it until now 😄

> Boundary overflow.

Exactly. Boundary check is needed.

Reply:
Hmm, good catch, it is valid. So the simple dumb check for `[]` in my code above is not right.
I have updated my example code above to not catch invalid `[]`.

I'm fine we only check for unpaired and out-of-ordered open/close brackets for now, but we do need to make the method internal until the behavior is finalized.

Reply:
Since `[[]` is also a valid pattern, I updated the code to not consider lone ']' as wildcard character. Please let me know if there is any problem.

Reply:
Seems It still doesn't work well. Try "[ [ ]".
I suggest to make the method public temporary, collect possible patterns and create Pester tests that help to design right algorithm. Before merge we make the method internal and hide/pending the tests.

Reply:
I tested the algorithm with a simple script:
```PowerShell
function ContainsValidWildcardPattern
{
    param (
        [string] $pattern
    )

    if ([string]::IsNullOrEmpty($pattern))
    {
        return $false;
    }

    $hasWildcard = $false;
    $openBracket = $false;

    for ($index = 0; $index -lt $pattern.Length; ++$index)
    {
        switch ($pattern[$index])
        {
            '*' {
                $hasWildcard = $true
            }

            '?' {
                $hasWildcard = $true
            }

            '[' {
                $hasWildcard = $true
                $openBracket = $true
            }

            ']' {
                # ']' is wildcard only if '[' exists
                # so we do not set hasWildcard here
                $openBracket = $false
            }

            # If it is an escape character then advance past
            # the next character
            '`' {
                ++$index
            }
        }
    }

    if ($openBracket) {
        return $false
    }

    return $hasWildcard
}

@(
    @{input = '['; IsPattern = $false}
    @{input = '[]'; IsPattern = $false}
    @{input = ']['; IsPattern = $false}
    @{input = '][]'; IsPattern = $false}
    @{input = '[[]'; IsPattern = $true}
    @{input = '[]]'; IsPattern = $true}
    @{input = '[][]'; IsPattern = $true}
    @{input = ']abc'; IsPattern = $false}
    @{input = '[]abc'; IsPattern = $false}
    @{input = '[abc[]abc]'; IsPattern = $true}
) | % {
    $out = if ($(ContainsValidWildcardPattern $_.input) -ne $_.IsPattern) {
        @{code = 31; verb = if ($_.IsPattern) { 'should be' } else { 'should not be' }}
    } else {
        @{code = 32; verb = if ($_.IsPattern) { 'is' } else { 'is not' }}
    }
    "`e[$($out.code)m'$($_.input)' $($out.verb) wildcard pattern`e[0m"
}
```
And the result is:
```
'[' is not wildcard pattern
'[]' should not be wildcard pattern
'][' is not wildcard pattern
'][]' should not be wildcard pattern
'[[]' is wildcard pattern
'[]]' is wildcard pattern
'[][]' is wildcard pattern
']abc' is not wildcard pattern
'[]abc' should not be wildcard pattern
'[abc[]abc]' is wildcard pattern
```
I will try to improve if this is still not good enough.